src/domain/resource/command.rs

```diff
@@ -110,6 +110,40 @@ pub async fn create(
     Ok(())
 }
 
+pub async fn update(
+    project_cache: Arc<dyn ProjectDrivenCache>,
+    resource_cache: Arc<dyn ResourceDrivenCache>,
+    event: Arc<dyn EventDrivenBridge>,
+    cmd: UpdateCmd,
+) -> Result<Resource> {
+    let Some(resource) = resource_cache.find_by_id(&cmd.id).await? else {
+        return Err(Error::CommandMalformed("invalid resource id".into()));
+    };
+
+    assert_permission(project_cache.clone(), &cmd.credential, &resource.project_id).await?;
+    let Some(project) = project_cache.find_by_id(&resource.project_id).await? else {
+        return Err(Error::CommandMalformed("invalid project id".into()));
+    };
+
```

Comment:
Add this `assert_project_resource(&project, &resource)?;` to prevent users from updating resources for other projects. This function will validate if the resource that is being updated is from the project_id.

Reply:
You can create a unit test to validate this scenario as well

Reply:
I can use that function if you want, but that check it being executed by the `assert_permission` function, isn't it? I am checking that the authenticated user has permission to do stuff on the resource's project

Reply:
`assert_permission` will just validate if the user has permission on that project, but the user can have permission in that project but use a different resource_id that is not from that project, because that the `assert_project_resource` is necessary. You can make a test creating 2 projects and 2 resources and try to update a resource from a different project_id. There is another way to do that which is passing the project_id for the `find_by_id` resource function, so the resource won't find.

Reply:
The RPC middleware auth will validate if the token is valid or secret but will not validate what the user wants to do and because that we need these kind of functions, so there are 3 levels of validation, first in the auth to validate the credentials, second to validate the project and the final to validate if that resource is from that project_id

Reply:
Maybe there is something I'm not seeing, but the project is extracted from the resource. If you check the lines before this line, I am doing the following:

* Find the resource in the cache (the id is sent on the `UpdateCmd`), if it doesn't exist return an error.
* Find the project from that resource, doing `project_cache.find_by_id(resource.project_id)`. So I already know that the resource corresponds to that project.
* Finally, assert that the user has permission on the project.

The project id comes from the resource, so I already know that the resource corresponds to the project. I can add the assert, but it won't do anything, because there is no way that the resource doesn't belong to the project (again, because the project is extracted from the resource and not the command)

Reply:
Yes, you are right, I didn't notice that you were getting the project_id from the resource cache.